src/kernel/include/thd.h

```diff
@@ -500,16 +497,106 @@ thd_invstk_push(struct thread *thd, struct comp_info *ci, unsigned long ip, unsi
 	prev->sp = sp;
 	memcpy(&top->comp_info, ci, sizeof(struct comp_info));
 	top->ip = top->sp = 0;
+	
+	if (likely(thd->ulk_invstk)) top->ulk_stkoff = thd->ulk_invstk->top;
 
 	return 0;
 }
 
+static inline int
+thd_invstk_push_ulk(struct thread *thd, struct comp_info *ci, struct cos_cpu_local_info *cos_info)
+{
+	struct invstk_entry *top;
+
+	if (unlikely(curr_invstk_top(cos_info) >= THD_INVSTK_MAXSZ)) return -1;
+
+	top  = &thd->invstk[curr_invstk_top(cos_info) + 1];
+	curr_invstk_inc(cos_info);
+	memcpy(&top->comp_info, ci, sizeof(struct comp_info));
+	
+	if (likely(thd->ulk_invstk)) top->ulk_stkoff = THD_UL_INV_ORIGIN;
+
+	return 0;
+}
+
+static inline void
+thd_invstk_pop_ulk(struct thread *thd, struct cos_cpu_local_info *cos_info)
+{
+	struct invstk_entry *curr;
+
+	if (unlikely(curr_invstk_top(cos_info) == 0)) return;
+	curr = &thd->invstk[curr_invstk_top(cos_info)-1];
+
+	if (curr->ulk_stkoff == THD_UL_INV_ORIGIN) {
+		curr_invstk_dec(cos_info);
+	}
+
+}
+
 static inline struct comp_info *
 thd_invstk_pop(struct thread *thd, unsigned long *ip, unsigned long *sp, struct cos_cpu_local_info *cos_info)
 {
+	struct invstk_entry *curr;
+	struct comp_info    *ci;
+
 	if (unlikely(curr_invstk_top(cos_info) == 0)) return NULL;
 	curr_invstk_dec(cos_info);
-	return thd_invstk_current(thd, ip, sp, cos_info);
+	curr = &thd->invstk[curr_invstk_top(cos_info)];
+
+	/* if we pushed over a UL inv entry, need to pop that too */
+	if (curr->ulk_stkoff == THD_UL_INV_ORIGIN) {
+		curr_invstk_dec(cos_info);
+	}
+
+	ci  = &curr->comp_info;
+	*ip = curr->ip;
+	*sp = curr->sp;
+
+	return ci;
+}
+
+/* defined in ulinv.c */
+struct comp_info *ulinvstk_current(struct ulk_invstk *stk, struct comp_info *origin, unsigned long offset);
+
+static inline struct comp_info *
+thd_invstk_current(struct thread *curr_thd, unsigned long *ip, unsigned long *sp, struct cos_cpu_local_info *cos_info)
+{
+	/* curr_thd should be the current thread! We are using cached invstk_top. */
+	struct invstk_entry *curr;
+	struct ulk_invstk   *ulk_invstk;
+	struct comp_info    *ci;
+
+	curr = &curr_thd->invstk[curr_invstk_top(cos_info)];
+	ulk_invstk = curr_thd->ulk_invstk;
+
+	/* this thread makes no UL-invs */
+	if (unlikely(!ulk_invstk)) {
+		*ip = curr->ip;
+		*sp = curr->sp;
+		return &curr->comp_info;
+	}
+
+	/* 
+	 * if there are new entries on the UL stack, 
+	 * we must be coming from a comp on the UL stack 
+	 */
+	if (ulk_invstk->top > curr->ulk_stkoff) {
```

Comment:
I'm missing how `ulk_stkoff` is passed to subsequent stack entries. Would need it if we had ULINV -> KINV -> *KINV* -> ULINV -> KINV. In the highlighted KINV, we'd need to continue tracking were we were in the ULINV from the past calculation, right?